moveit_py/src/moveit/moveit_py_utils/src/copy_ros_msg.cpp

```diff
@@ -0,0 +1,362 @@
+  orientation_constraint_cpp.absolute_x_axis_tolerance =
+      orientation_constraint.attr("absolute_x_axis_tolerance").cast<double>();
+  orientation_constraint_cpp.absolute_y_axis_tolerance =
+      orientation_constraint.attr("absolute_y_axis_tolerance").cast<double>();
+  orientation_constraint_cpp.absolute_z_axis_tolerance =
+      orientation_constraint.attr("absolute_z_axis_tolerance").cast<double>();
+  orientation_constraint_cpp.parameterization = orientation_constraint.attr("parameterization").cast<int>();
+  orientation_constraint_cpp.weight = orientation_constraint.attr("weight").cast<double>();
+
+  return orientation_constraint_cpp;
+}
+
+moveit_msgs::msg::VisibilityConstraint VisibilityConstraintToCpp(const py::object& visibility_constraint)
+{
+  // recreate instance in C++ using python object data
+  moveit_msgs::msg::VisibilityConstraint visibility_constraint_cpp;
+  visibility_constraint_cpp.target_radius = visibility_constraint.attr("target_radius").cast<double>();
+  visibility_constraint_cpp.target_pose = PoseStampedToCpp(visibility_constraint.attr("target_pose"));
+  visibility_constraint_cpp.cone_sides = visibility_constraint.attr("cone_sides").cast<int>();
+  visibility_constraint_cpp.sensor_pose = PoseStampedToCpp(visibility_constraint.attr("sensor_pose"));
+  visibility_constraint_cpp.max_view_angle = visibility_constraint.attr("max_view_angle").cast<double>();
+  visibility_constraint_cpp.max_range_angle = visibility_constraint.attr("max_range_angle").cast<double>();
+  visibility_constraint_cpp.sensor_view_direction = visibility_constraint.attr("sensor_view_direction").cast<int>();
+  visibility_constraint_cpp.weight = visibility_constraint.attr("weight").cast<double>();
+
+  return visibility_constraint_cpp;
+}
+
+moveit_msgs::msg::CollisionObject CollisionObjectToCpp(const py::object& collision_object)
+{
+  //  recreate instance in C++ using python object data
+  moveit_msgs::msg::CollisionObject collision_object_cpp;
+
+  // header
+  collision_object_cpp.header.frame_id = collision_object.attr("header").attr("frame_id").cast<std::string>();
+
+  // object pose
+  collision_object_cpp.pose = PoseToCpp(collision_object.attr("pose"));
+
+  // object id
+  collision_object_cpp.id = collision_object.attr("id").cast<std::string>();
+
+  // object type
+  collision_object_cpp.type.key = collision_object.attr("type").attr("key").cast<std::string>();
+  collision_object_cpp.type.db = collision_object.attr("type").attr("db").cast<std::string>();
+
+  // iterate through python list creating C++ vector of primitives
+  int num_primitives = collision_object.attr("primitives").attr("__len__")().cast<int>();
+  for (int i = 0; i < num_primitives; i++)
+  {
+    py::object primitive = collision_object.attr("primitives").attr("__getitem__")(i);
+    auto primitive_cpp = SolidPrimitiveToCpp(primitive);
+    collision_object_cpp.primitives.push_back(primitive_cpp);
+  }
```

Comment:
For this and many other equivalent loops, is it possible to use range-based loops instead? So e.g.,

```suggestion
  for (const auto& primitive : collision_object.attr("primitives"))
  {
    auto primitive_cpp = SolidPrimitiveToCpp(primitive);
    collision_object_cpp.primitives.push_back(primitive_cpp);
  }
```

... or is there some pybind11 limitation?

Reply:
Issue: https://github.com/ros-planning/moveit2/issues/1930.

Reply:
Not that the commit history here is clean :) but it might make sense to address this enhancement all at once in a single issue.